app/service/form_data.go

```diff
@@ -0,0 +1,214 @@
+func (f *FormData) ConstructMapForDB() map[string]interface{} {
+	result := map[string]interface{}{}
+	f.addDBFieldsIntoMap(result, reflect.ValueOf(f.definitionStructure), "")
+	return result
+}
+
+var mapstructTypeErrorRegexp = regexp.MustCompile(`^'([^']*)'\s+(.*)$`)
+var mapstructDecodingErrorRegexp = regexp.MustCompile(`^error decoding '([^']*)':\s+(.*)$`)
+
+func (f *FormData) decodeRequestJSONDataIntoStruct(r *http.Request) error {
+	var rawData map[string]interface{}
+	if err := json.NewDecoder(r.Body).Decode(&rawData); err != nil {
+		return err
+	}
+
+	f.fieldErrors = make(FieldErrors)
+	f.usedKeys = map[string]bool{}
+
+	decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		Result:           f.definitionStructure,
+		DecodeHook:       mapstructure.StringToTimeHookFunc(time.RFC3339),
+		ErrorUnused:      false, // we will check this on our own
+		Metadata:         &f.metadata,
+		TagName:          "json",
+		ZeroFields:       true, // this marks keys with null values as used
+		WeaklyTypedInput: false,
+	})
+
+	if err != nil {
+		panic(err) // this error can only be caused by bugs in the code
+	}
+
+	if err := decoder.Decode(rawData); err != nil {
+		if err, ok := err.(*mapstructure.Error); ok {
+			for _, fieldErrorString := range err.Errors { // Convert mapstructure's errors to our format
+				if matches := mapstructTypeErrorRegexp.FindStringSubmatch(fieldErrorString); len(matches) > 0 {
+					key := make([]byte, len(matches[1]))
+					copy(key, matches[1])
+					value := make([]byte, len(matches[2]))
+					copy(value, matches[2])
+					f.fieldErrors[string(key)] = append(f.fieldErrors[string(key)], string(value))
+					continue
+				}
+				if matches := mapstructDecodingErrorRegexp.FindStringSubmatch(fieldErrorString); len(matches) > 0 {
+					key := make([]byte, len(matches[1]))
+					copy(key, matches[1])
+					f.fieldErrors[string(key)] = append(f.fieldErrors[string(key)], "decoding error: "+matches[2])
+					continue
+				}
+				f.fieldErrors[""] = append(f.fieldErrors[""], fieldErrorString) // should never happen
+			}
+		}
+	}
+	return nil
+}
+
+func (f *FormData) validateFieldValues() error {
+	if _, err := govalidator.ValidateStruct(f.definitionStructure); err != nil {
+		if unhandledErr := f.processGovalidatorErrors(err); unhandledErr != nil {
+			return err // should never happen
+		}
+	}
+	return nil
+}
+
+func (f *FormData) processGovalidatorErrors(err error) error {
+	if validatorErrors, ok := err.(govalidator.Errors); ok {
+		for _, validatorError := range validatorErrors {
+			if err, ok := validatorError.(govalidator.Error); ok {
+				path := strings.Join(err.Path, ".")
+				if len(path) > 0 {
+					path += "."
+				}
+				path += err.Name
+				f.fieldErrors[path] = append(f.fieldErrors[path], err.Err.Error())
+			} else {
+				if err := f.processGovalidatorErrors(validatorError); err != nil {
+					return err // should never happen
+				}
+			}
+		}
+	} else {
+		return err // should never happen
```

Comment:
If it should not happens and panic anyway at the end, shouldn't you use the type assertion without the `ok` directly? I would say it would be a more "Go-ish" approach... and it would decrease the not-tested-as-not-testable code.

Reply:
fixed